### PR TITLE
http: flip runtime flag to disable route config searching downgrading

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: http: flip runtime flag to disable route config searching downgrading
Additional Description: 

According to previous discussion, flip the runtime flag to true after Q4 release.

Risk Level: low.
Testing: n/a.
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

**Comment by wbpcode** [comment]
The biggest impact of this PR is that the embedded filter of composite filter could not use the canonical filter name (e.g. `envoy.filters.http.buffer`) to get route level per filter config. And because the filter chain could only get the config name of composite filter itself, so all embedded filters in same composite filter will share same config name in current implementation, which may be unexpected.

A possible solution is to create a wapper of `StreamFilterCallbacks` to override `mostSpecificPerFilterConfig` method. Then the composite set this wrapper into its embedded filters. By this way, every embedded filter could use its own config name to get route level per filter config.

cc @tyxia @snowp

**Comment by tyxia** [comment]
@wbpcode Thanks for letting us know. Composite filter is marked as stable and actively used in our released product. 

So I think the desired behavior is implementing the related solution first before checking in breaking change.

**Comment by tyxia** [comment]
/assign @tyxia 

I will take a closer look at potential impact of composite filter and update here.

**Comment by tyxia** [comment]
Thanks for waiting on this! @wbpcode 

First of all, I am fully onboard with deprecating filter canonical name. That is just an opaque name which can be arbitrarily set in FactoryBase c'tor.  Keeping single source of truth is cleaner and more robust. 

Second, your change will NOT break our product's use case because we have noticed this and always stuck with with filter config name, more specifically top-level filter config name. cc @htuch and @rshriram 

Third, we do have an internal tracking bug to surface embedded filter's config name (not canonical name). We should have cycle to work on this in Q4 (cc @jstraceski) 

Last, AFAIU your change only breaks the usage of canonical name `envoy.filters.http.match_delegate` as that is the top-level filter plugged in the filter chain. i.e., those embedded filter's canonical name is already not visible even in that fallback name.

I will add a integration test in composite filter integration test to validate the expected behavior.

**Comment by wbpcode** [comment]
> Last, AFAIU your change only breaks the usage of canonical name envoy.filters.http.match_delegate as that is the top-level filter plugged in the filter chain. i.e., those embedded filter's canonical name is already not visible even in that fallback name.

Right, but anyway, those canonical name is already unavailable, this is more like a inherent problem of match_delegate. And at least, this PR won't make thing worse.

**Comment by wbpcode** [comment]
I am not in a hurry. So, cc @tyxia could you give a stamp when you think this PR is OK to you?

File: test/extensions/filters/http/lua/lua_integration_test.cc [hunk 1]
```diff
@@ -908,7 +908,7 @@ name: basic_lua_routes
     route:
       cluster: lua_cluster
     typed_per_filter_config:
-      envoy.filters.http.lua:
+      lua:
```
Comment (tyxia) [inline test/extensions/filters/http/lua/lua_integration_test.cc:911]: `envoy.filters.http.lua` will still be an valid key as long as it is also specified in [filter config name](https://github.com/envoyproxy/envoy/blob/65b1f6039236e466c196e3cfbe082567c569b640/envoy/http/filter_factory.h#L33), right?

In other words, there is no favoring of  `lua` over `envoy.filters.http.lua` from naming perspective?

**Review by tyxia — APPROVED** [review]
LGTM, thanks!

**Review by alyssawilk — COMMENTED** [review]
minor nits otherwise lg

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**